lib/matplotlib/lines.py

```diff
@@ -1537,27 +1536,29 @@ def process_selected(self, ind, xs, ys):
 
         selector = HighlightSelected(line)
         plt.show()
-
     """
+
     def __init__(self, line):
         """
-        Initialize the class with a `.Line2D`.  The line should already be
-        added to an `~.axes.Axes` and should have the picker property set.
+        Parameters
+        ----------
+        line : `.Line2D`
+            The line must already have been added to an `~.axes.Axes` and must
+            have its picker property set.
         """
         if line.axes is None:
             raise RuntimeError('You must first add the line to the Axes')
-
         if line.get_picker() is None:
             raise RuntimeError('You must first set the picker property '
                                'of the line')
-
         self.axes = line.axes
         self.line = line
-        self.canvas = self.axes.figure.canvas
-        self.cid = self.canvas.mpl_connect('pick_event', self.onpick)
-
+        self.cid = self.canvas.callbacks._connect_picklable(
+            'pick_event', self.onpick)
         self.ind = set()
 
+    canvas = property(lambda self: self.axes.figure.canvas)
```

Comment:
Was there a reason we needed `canvas` in the first place?

Reply:
I would not have made this public to start with, but e.g. the HighlightSelector example (in the docstring) directly accesses it and it is marginally more convenient to have it there, so let's not bother with deprecating it?